apollo-runtime/src/main/java/com/apollographql/apollo/subscription/AppSyncOperationMessageSerializer.kt

```diff
@@ -0,0 +1,81 @@
+package com.apollographql.apollo.subscription
+
+import com.apollographql.apollo.api.internal.json.JsonWriter
+import com.apollographql.apollo.api.internal.json.Utils
+import com.apollographql.apollo.api.internal.json.writeObject
+import com.apollographql.apollo.subscription.ApolloOperationMessageSerializer.writePayloadContentsTo
+import okhttp3.HttpUrl
+import okio.Buffer
+import okio.BufferedSink
+import java.util.Base64
+
+class AppSyncOperationMessageSerializer(
+    private val authorization: Map<String, Any?>
+) : OperationMessageSerializer by ApolloOperationMessageSerializer {
+  override fun OperationClientMessage.writeTo(sink: BufferedSink) {
+    when (this) {
+      is OperationClientMessage.Start -> JsonWriter.of(sink).use { writeTo(it) }
+      is OperationClientMessage.Init,
+      is OperationClientMessage.Stop,
+      is OperationClientMessage.Terminate -> with(ApolloOperationMessageSerializer) { writeTo(sink) }
+    }
+  }
+
+  private fun OperationClientMessage.Start.writeTo(writer: JsonWriter) {
+    writer.writeObject {
+      name(ApolloOperationMessageSerializer.JSON_KEY_ID).value(subscriptionId)
+      name(ApolloOperationMessageSerializer.JSON_KEY_TYPE).value(OperationClientMessage.Start.TYPE)
+      name(ApolloOperationMessageSerializer.JSON_KEY_PAYLOAD).writeObject {
+        name(JSON_KEY_DATA).writeObject {
+          writePayloadContentsTo(writer)
+        }
+
+        if (autoPersistSubscription) {
+          name(ApolloOperationMessageSerializer.JSON_KEY_EXTENSIONS).writeObject {
+            name("authorization")
+            Utils.writeToJson(authorization, writer)
+          }
+        }
+      }
+    }
+  }
+
+  companion object {
+    private const val JSON_KEY_DATA = "data"
+
+    @JvmOverloads
+    @JvmStatic
+    fun buildAppSyncWebSocketUrl(
+        baseWebSocketUrl: String,
+        authorization: Map<String, Any?>,
+        payload: Map<String, Any?> = emptyMap(),
+    ): String =
+        baseWebSocketUrl
+          .let { url ->
+            when {
+              url.startsWith("ws://", ignoreCase = true) -> "http" + url.drop(2)
+              url.startsWith("wss://", ignoreCase = true) -> "https" + url.drop(3)
+              else -> url
+            }
+          }
+          .let { HttpUrl.get(it) }
+          .newBuilder()
+          .setQueryParameter("header", authorization.base64Encode())
+          .setQueryParameter("payload", payload.base64Encode())
+          .build()
+          .toString()
+          .let { url ->
+            when {
+              url.startsWith("http://", ignoreCase = true) -> "ws" + url.drop(4)
+              url.startsWith("https://", ignoreCase = true) -> "wss" + url.drop(5)
+              else -> url
```

Comment:
Ouch. I guess we want to keep the `ws://` to avoid confusing any user there, even if ultimately, okhttp is going to replace everything with `http://` ? That feels a bit awkward but I don't really have another suggestion besides using `java.net.URL` that will end up equally verbose.

Reply:
Yeah, that's what I thought too. `URL` doesn't help here because it doesn't have any builder pattern.